segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/TableIterator.java

```diff
@@ -0,0 +1,326 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.server.tables;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Maps;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.common.util.AsyncIterator;
+import io.pravega.segmentstore.contracts.Attributes;
+import io.pravega.segmentstore.server.AttributeIterator;
+import io.pravega.segmentstore.server.DirectSegmentAccess;
+import java.time.Duration;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import javax.annotation.concurrent.ThreadSafe;
+import lombok.AccessLevel;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.val;
+
+/**
+ * Iterates through {@link TableBucket}s in a Segment.
+ * @param <T> Type of the final, converted result.
+ */
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+@ThreadSafe
+class TableIterator<T> implements AsyncIterator<T> {
+    //region Members
+
+    private final AttributeIterator indexHashIterator;
+    private final ConvertResult<T> resultConverter;
+    private final ArrayDeque<Map.Entry<UUID, Long>> cacheHashes;
+    private final Executor executor;
+    private final AtomicReference<Iterator<TableBucket>> currentBatch = new AtomicReference<>();
+    private final AtomicBoolean inProgress = new AtomicBoolean(false);
+
+    //endregion
+
+    //region AsyncIterator Implementation
+
+    @Override
+    public CompletableFuture<T> getNext() {
+        // Verify no other call to getNext() is currently executing.
+        Preconditions.checkState(this.inProgress.compareAndSet(false, true), "Another call to getNext() is in progress.");
```

Comment:
Why do you prefer to error here rather than synchronize the computation and serialize the execution of calls? The processor calling this will have to know that it needs to handle runtime exceptions, even though they haven't done anything wrong.